### PR TITLE
Mohavro patch 12 2

### PR DESCRIPTION
**Pull request by mohavro**
## Description

Updating branch mohavro-12 with latest commits from previously merged branch mohavro-patch-12-2.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test addition/update
- [ ] Dependency update

## Related Issues

Closes #

## Changes Made

-
-
-

## Testing

- [ ] Unit tests pass locally
- [ ] Integration tests pass locally
- [ ] Manual testing completed
- [ ] New tests added for new functionality

### Test Commands

```bash
# Commands used to test
pytest
npm test
```

### Code Quality

- [ ] My code follows the project's style guidelines
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Documentation

- [ ] I have updated the README.md (if needed)
- [ ] I have updated the CHANGELOG.md
- [ ] I have updated relevant documentation files
- [ ] I have added/updated docstrings

### Dependencies

- [ ] I have checked for dependency conflicts
- [ ] I have updated requirements.txt (if needed)
- [ ] I have updated package.json (if needed)

### Branch Management

- [ ] My branch is up to date with main
- [ ] I will delete this branch after merge
- [ ] This branch has a descriptive name
- [ ] This PR has been open for less than 2 weeks

## Additional Notes

**For Reviewers**: Please ensure the branch is deleted after merging this PR to maintain repository hygiene.

**Related Documentation**:
- [Contributing Guidelines](../CONTRIBUTING.md)
- [Branch Cleanup Guidelines](../BRANCH_CLEANUP_QUICK_REFERENCE.md)

---
## Summary by cubic
Refactors 2D/3D graph visuals and API/DB utilities, rewrites the schema report CLI, and expands tests and CI to improve safety, clarity, and maintainability. Adds strict visualization input validation, moves session_scope to the repository, and hardens Bandit/CI configs.

- **Refactors**
  - 2D/3D visuals: stronger input checks (IDs, colors), clearer trace separation and directional arrows; embeds reports via Plotly layout.meta.
  - CLI: schema_report_cli rebuilt with Typer to generate MD/HTML and save outputs; simpler load_graph helpers.
  - DB/API/Graph: session_scope moved to src/data/repository and re-exported; AssetRelationshipGraph cleanup; CORS origin validator and lifespan tweaks; formulaic analysis/docstring fixes; sample data uses corporate_link relationship.

- **CI/Tooling/Tests**
  - Security/CI: Bandit now excludes tests; CI workflow cleaned; Codacy runtimes updated; .gitignore adds .qlty; dev deps switch to PyYAML>=6 and add ruff.
  - Tests: new suites for Gradio app, suggest_fixes, generate_status, and schema_report_cli; expanded DB/session, API/CORS, repository coverage; adjusted markers and stricter redaction.

<sup>Written for commit 52660955df2429e0dde450d45613d5a5e4190200. Summary will update on new commits. <a href="https://cubic.dev/pr/DashFin-FarDb/financial-asset-relationship-db/pull/863">Review in cubic</a></sup>

### **Description**
- Introduced extensive unit tests for `AssetGraphRepository`, covering CRUD operations, relationships, and regulatory events.
- Expanded database configuration tests to ensure robust error handling and transaction management.
- Improved test coverage for edge cases and concurrent access scenarios.

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_repository.py</strong><dd><code>Comprehensive Unit Tests for AssetGraphRepository</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_repository.py
<li>Added comprehensive unit tests for AssetGraphRepository.<br> <li> Included tests for asset CRUD operations, relationship management, and <br>regulatory event handling.<br> <li> Enhanced data transformation and edge case tests.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/863/files#diff-ec0158fb38d734ff52212c671b2bc4425a3f4d20a6add21119c420556d0a65f4">+1562/-1266</a></td>
</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_database.py</strong><dd><code>Enhanced Unit Tests for Database Configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/test_database.py
<li>Expanded unit tests for database configuration helpers.<br> <li> Added tests for engine creation, session factory, and database <br>initialization.<br> <li> Included transaction scope management and error handling tests.<br>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/863/files#diff-1dac77ff25a35547d1f7050c0644d561ce508dddaad1187de0567e7f9d821d4b">+699/-399</a></td>
</tr>
</table></td></tr></tr></tbody></table>

> 💡 **Penify usage**:
>Comment `/help` on the PR to get a list of all available Penify tools and their descriptions

___

## **CodeAnt-AI Description**
Validate inputs and make 3D asset network visualizations safer and clearer; add tests and harden PR suggestion script

### What Changed
- 3D graph visualization now validates all inputs (positions, asset IDs, colors, hover texts) and returns clear errors instead of failing silently; figures include consistent dynamic titles, styled node markers, separate relationship traces, and optional directional arrows.
- Relationship processing was made robust and performant: relationships are indexed with thread-safety checks, grouped by type/direction, and rendered as distinct traces so filters and arrow toggles reliably control what users see.
- The PR suggestions script was hardened: missing dependency messages are clearer, code-suggestion extraction and comment parsing handle more patterns, and output writing to CI summary and temp files includes safer error handling.
- Large set of unit tests added/updated: comprehensive tests for the suggestion script, new validation/unit tests for review parsing, and adjusted expectations in real-data fetching tests (reduced expected ticker calls).

### Impact
`✅ Clearer visualization errors`
`✅ Fewer visualization crashes when given bad data`
`✅ More reliable actionable suggestion reports`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: src/visualizations/graph_2d_visuals.py [hunk 1]
```diff
@@ -83,7 +251,68 @@ def visualize_2d_graph(
     ):
         fig.add_trace(trace)
 
-    fig.add_trace(_create_node_trace(graph, positions, asset_ids))
+    # Add node trace
+    node_x = [positions[asset_id][0] for asset_id in asset_ids]
+    node_y = [positions[asset_id][1] for asset_id in asset_ids]
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/visualizations/graph_2d_visuals.py]:
```suggestion
    # Guard: some assets may be absent from positions (e.g. no relationships)
    positioned_ids = [aid for aid in asset_ids if aid in positions]
    node_x = [positions[aid][0] for aid in positioned_ids]
    node_y = [positions[aid][1] for aid in positioned_ids]
```

File: tests/unit/test_app_gradio.py [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+
+    @patch("app.real_data_fetcher")
+    def test_create_database_tries_multiple_candidates(self, mock_fetcher):
+        """Test that _create_database tries multiple function candidates."""
+        mock_graph = MagicMock()
+        mock_fetcher.create_real_database = Mock(return_value=mock_graph)
+
+        result = FinancialAssetApp._create_database()
+
+        assert result == mock_graph
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_app_gradio.py]:
```suggestion
    @patch("app.real_data_fetcher")
    def test_create_database_tries_multiple_candidates(self, mock_fetcher):
        """Test that _create_database tries multiple function candidates."""
        from src.logic.asset_graph import AssetRelationshipGraph
        mock_graph = MagicMock(spec=AssetRelationshipGraph)
        mock_fetcher.create_real_database = Mock(return_value=mock_graph)

        result = FinancialAssetApp._create_database()

        assert result == mock_graph
```

File: .qlty/plugin_cachedir [hunk 3]
```diff
@@ -0,0 +1 @@
+C:/Users/Administrator/.qlty/cache/repos/1b4160912774202b565449fbd687987c/plugin_cachedir
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .qlty/plugin_cachedir]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/863#discussion_r2835931755)

File: tests/unit/test_summary_documentation.py [hunk 4]
```diff
@@ -1,236 +1,240 @@
...
+@pytest.mark.unit
+class TestEnhancedTestSummary:
+    """Test cases for ENHANCED_TEST_SUMMARY.md."""
+
+    @pytest.fixture
+    def summary_path(self):
+        """
+        Provide the path to the enhanced test summary file.
+
+        Returns:
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/unit/test_summary_documentation.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/863#discussion_r2835931758)

**Review by mohavro — COMMENTED** [review]
Approved
